definitions/npm/react-apollo_v2.x.x/flow_v0.58.x-/react-apollo_v2.x.x.js

```diff
@@ -276,25 +276,38 @@ declare module 'react-apollo' {
 
   declare export class Subscription<TData> extends React$Component<SubscriptionProps<TData>> {}
 
-  declare export type MutationFunction<TVariables> = (options: {
+  declare type ExecutionResult<T> = {
+    data?: T;
+    extensions?: {[string]: any};
+    errors?: any[];
+  }
+
+  declare type FetchResult<C={[string]: any}, E={[string]: any}> = ExecutionResult<C> & {
+    extensions?: E,
+    context?: C
+  }
+
+  declare type OperationVariables = {[string]: any}
+
+  declare export type MutationFunction<TData=any, TVariables=OperationVariables> = (options: {
     variables?: TVariables,
     optimisticResponse?: Object,
-    refetchQueries?: (mutationResult: FetchResult) => Array<{query: DocumentNode, variables: {[string]: any}}>,
-    update?: (cache: DataProxy, mutationResult: FetchResult) => any
-  }) => Promise<*>
+    refetchQueries?: (mutationResult: FetchResult<>) => Array<{query: DocumentNode, variables: {[string]: any}}>,
+    update?: (cache: DataProxy, mutationResult: FetchResult<>) => any
+  }) => Promise<void | FetchResult<TData>>
 
-  declare export type MutationResult<TData> = {loading: boolean, error?: ApolloError, data?: TData, called: boolean}
+  declare export type MutationResult<TData={[string]: any}> = {loading: boolean, error?: ApolloError, data?: TData, called: boolean}
 
-  declare export type MutationRenderPropFunction<TData, TVariables> = (mutate: MutationFunction<TVariables>, result: MutationResult<TData>) => React$Node
+  declare export type MutationRenderPropFunction<TData, TVariables> = (mutate: MutationFunction<TData, TVariables>, result: MutationResult<TData>) => React$Node
 
   declare export class Mutation<TData, TVariables=void> extends React$Component<{
     mutation: DocumentNode,
     children: MutationRenderPropFunction<TData, TVariables>,
     variables?: TVariables,
-    update?: (cache: DataProxy, mutationResult: FetchResult) => any,
+    update?: (cache: DataProxy, mutationResult: FetchResult<>) => any,
     ignoreResults?: boolean,
     optimisticResponse?: Object,
-    refetchQueries?: (mutationResult: FetchResult) => Array<{query: DocumentNode, variables: TVariables}>,
+    refetchQueries?: string[] | {query: DocumentNode, variables?: {[string]: any}} | ((mutationResult: FetchResult<>) => Array<{query: DocumentNode, variables?: {[string]: any}}>),
```

Comment:
https://github.com/apollographql/react-apollo/blob/master/src/types.ts#L22 The function can actually return an array of strings or an array of PureQueryOptions. Maybe it would be good to define RefetchQueriesProviderFnand PureQueryOptions as well to make a little less crowded.